### PR TITLE
Remove hand crafted types for text-encoding

### PR DESCRIPTION
**Pull request by mohsen1**
Still WIP but comments are welcome 

This change:

* Removes hand written types for `text-encoding` module and uses `@types/text-encoding`
* Updates the list script to enforce TypeScript type checking and Prettier code lint (must ran Prettier before) 

TODO List:

* [x] Resolve all tsc issues 
  * [x] Fix duplicate declaration for global `TextDecoder` and `TextEncoder`. It already exist in `dom.d.ts`
  * [x] Figure out typing for `*.pb` (protobuf?) files. I need help with this. Should we have a declaration per file? Can we use something like [Proto2TypeScript](https://github.com/SINTEF-9012/Proto2TypeScript) to automate it?

### CODE REVIEW

File: declarations.d.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
+declare module "*.pb" {
```
Comment (qti3e) [inline declarations.d.ts]: protobuf auto-generates this file based on contents inside `msg.proto`, look at `msg.pb.d.ts` after running:
```
make deno
```
or in case you don't want to build v8 (which takes ~30min)
```
make msg.pb.d.ts

**Comment by ry** [comment]
@mohsen1 thanks! I applaud the effort. If you can remove text-encoding.d.ts while also not depending on @types/text-encoding, I would be very happy. I tried a bit to do this but got stuck.

File: .gitignore [hunk 2]
```diff
@@ -6,3 +6,4 @@ assets.go
 msg.pb.go
 msg.pb.js
 msg.pb.d.ts
+.vscode
```
[the remark this answers is not in the document]
Comment (mohsen1) [reply .gitignore]: That's nice and I did it on my machine. I think it's good to have it here too since VSCode is pretty popular these days :)

[the remark this answers is not in the document]
Comment (fearphage) [reply .gitignore]: I find it it best to keep the files in each repository specific to that project instead of about the people editing it. This project has no reliance on or interactions with VS Code. But it all comes down to personal preference in the end.

**Comment by mohsen1** [comment]
@ry you can't import from `"text-encoding"` module without having it declared somewhere. 

https://github.com/ry/deno/blob/a4d89f2ee1ddde786f2335eada86f00e3efd1174/globals.ts#L54

This is the minimal declaration if you don't like to have a dependency just for types.

File: tsconfig.json [hunk 3]
```diff
@@ -1,14 +1,13 @@
 {
   "compilerOptions": {
-    "allowJs": true,
```
Comment (yorkie) [inline tsconfig.json]: Why removing this option?

Comment (mohsen1) [reply tsconfig.json]: shoot! I didn't meant to dod that!

File: .gitignore (hunk 2 above)
Comment (qti3e) [inline .gitignore]: Unrelated change

Comment (qti3e) [inline .gitignore]: Unrelated change

File: package.json [hunk 4]
```diff
@@ -1,8 +1,8 @@
 {
   "name": "deno",
   "scripts": {
-    "lint": "tslint -p tsconfig.json -c tslint.json",
-    "fmt": "prettier --write *.ts* *.js *.json"
+    "lint": "tslint -p . && prettier --list-different \"{*.ts,js,json}\"",
+    "fmt": "prettier --write \"{*.ts,js,json}\""
```
Comment (qti3e) [inline package.json]: Unrelated change

File: runtime.ts [hunk 5]
```diff
@@ -149,7 +149,7 @@ export function resolveModule(
   let fetchResponse;
   try {
     fetchResponse = os.codeFetch(moduleSpecifier, containingFile);
-  } catch(e) {
+  } catch (e) {
```
[the remark this answers is not in the document]
Comment (qti3e) [reply runtime.ts]: Unrelated change

**Comment by ry** [comment]
This doesn't solve the TODO. Thanks for researching it. We can revisit later with that TS issue is fixed.

**Comment by mohsen1** [comment]
I don't understand how it would be possible to import from text-encoding module if you don't want to depend on`@types/text-encoding` and delete text-encoding.d.ts. Module has to be declared somehow

**Comment by ry** [comment]
It's inside TS: https://github.com/Microsoft/TypeScript/blob/f8503f26328beb8755792528dd312f29f9d5415c/lib/lib.dom.d.ts#L13139-L13142

**Comment by mohsen1** [comment]
A module with "text-encoding" specifier is not defined in lib.dom.d.ts. If I'm reading the code correctly you are importing a runtime value from the text encoding module and assigning it to global object.

**Comment by ry** [comment]
Right.. so somehow we need to import the implementation (polyfill) but not have typescript complain.

**Comment by mohsen1** [comment]
To import those values from "text-encoding" module there should be a module declaration somewhere. There are two options :

1) create a minimal module declaration that declares exports of `TextEncoder` and `TextDecoder` with types of dom lib. That would be this PR. TypeScript has two variable namespaces. One for values and one for types. That's why I have `TextDecoder: TextDecoder`

2) depend on `@types/text-encoder` which declares the module. I'm trying to fix it to use dom types here https://github.com/DefinitelyTyped/DefinitelyTyped/pull/26141

**Comment by ry** [comment]
@mohsen1 oh gosh I'm very sorry. I miss read the commit - I thought it was still importing `@types/text-encoder`.
Thank you for solving this!
